src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs

```diff
@@ -177,12 +179,21 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
                 {
                     Error(diagnostics, ErrorCode.ERR_AttributeCtorInParameter, node, attributeConstructor.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat));
                 }
+
+                boundConstructorArguments = BuildArgumentsForErrorRecovery(constructorArguments, ImmutableArray.Create(attributeConstructor));
+                constructorArguments.Arguments.Free();
+            }
+            else
+            {
+                boundConstructorArguments = constructorArguments.Arguments.SelectAsArrayInPlaceAndFree(
+                    (arg, @this) => @this.BindToTypeForErrorRecovery(arg), this);
             }
 
-            var constructorArguments = analyzedArguments.ConstructorArguments;
-            ImmutableArray<BoundExpression> boundConstructorArguments = constructorArguments.Arguments.ToImmutableAndFree();
             ImmutableArray<string> boundConstructorArgumentNamesOpt = constructorArguments.GetNames();
-            ImmutableArray<BoundExpression> boundNamedArguments = analyzedArguments.NamedArguments;
+            ImmutableArray<BoundExpression> boundNamedArguments =
+                analyzedArguments.NamedArguments?.SelectAsArrayInPlaceAndFree((namedArg, @this) => @this.BindToTypeForErrorRecovery(namedArg), this)
```

Comment:
>@this.BindToTypeForErrorRecovery(namedArg) [](start = 99, length = 42)

Similar comment here. It doesn't feel appropriate to do this for success cases.

Reply:
I've looked into this, and we actually don't need to do this at all. These are always `BoundAssignmentOperator` nodes, and the right sides are always appropriately converted by binding. I've changed the code to strongly assume that these are always `BoundAssignmentOperator`s and simply assert that the `Right` never needs binding here.

---
In reply to: [577678797](https://github.com/dotnet/roslyn/pull/51206#discussion_r577678797) [](ancestors = 577678797)